File: src/main/java/com/lsd/diagram/SequenceDiagramGenerator.java
package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static org.apache.commons.collections4.ListUtils.partition;

@Builder
public class SequenceDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Optional<Diagram> diagram(int maxEventsPerDiagram) {
        if (events.isEmpty()) {
            return Optional.empty();
        }
        var uml = partition(events, maxEventsPerDiagram).stream()
                .map(this::generateSequenceUml)
                .collect(joining(System.lineSeparator()));

        String svg = generateSequenceSvg(uml);
        return Optional.of(Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(svg)
                .build());
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml(List<SequenceEvent> events) {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants.stream().distinct().collect(toList()),
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}

File: src/main/kotlin/com/lsd/diagram/SequenceDiagramGenerator.kt
package com.lsd.diagram

import com.github.jknack.handlebars.Handlebars
import com.lsd.IdGenerator
import com.lsd.builders.SequenceDiagramGeneratorBuilder
import com.lsd.events.SequenceEvent
import com.lsd.events.groupedByPages
import com.lsd.properties.LsdProperties
import com.lsd.properties.LsdProperties.DIAGRAM_THEME
import com.lsd.report.model.Diagram
import com.lsd.report.model.Participant
import org.apache.commons.collections4.ListUtils
import java.lang.System.lineSeparator
import java.util.stream.Collectors.joining
import java.util.stream.Collectors.toList

data class SequenceDiagramGenerator(
    private val idGenerator: IdGenerator,
    private val events: List<SequenceEvent> = emptyList(),
    private val includes: List<String> = emptyList(),
    private val participants: List<Participant> = emptyList(),
) {
    private val template = Handlebars().compile("templates/sequence-uml")

    fun diagram(maxEventsPerDiagram: Int): Diagram? {
        val uml = generateUml(maxEventsPerDiagram)
        return if (events.isEmpty()) null else {
            Diagram(
                id = idGenerator.next(),
                uml = uml,
                svg = convertToSvg(markup = uml)
            )
        }
    }

    private fun generateUml(maxEventsPerDiagram: Int): String =
        events.groupedByPages()
            .stream()
            .flatMap { ListUtils.partition(it, maxEventsPerDiagram).stream() }
            .map { generateSequenceUml(it) }
            .collect(joining(lineSeparator()))

    private fun generateSequenceUml(events: List<SequenceEvent>): String {
        return template.apply(
            mapOf(
                "theme" to LsdProperties.get(DIAGRAM_THEME),
                "includes" to includes.distinct(),
                "participants" to participants.stream()
                    .distinct()
                    .collect(toList()),
                "events" to events.stream()
                    .map(SequenceEvent::toMarkup)
                    .collect(toList())
            )
        )
    }

    companion object {
        fun builder(): SequenceDiagramGeneratorBuilder = SequenceDiagramGeneratorBuilder()
    }
}

